mapper/tedge_json/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+
+    #[error("Empty Thin Edge measurement: {name:?} must contain at least one measurement")]
+    EmptyThinEdgeJson { name: String },
+
+    #[error("Not a number: the {name:?} value must be a number, not {actual_type}.")]
+    InvalidThinEdgeJsonValue { name: String, actual_type: String },
+
+    #[error("Not a timestamp: the time value must be an ISO8601 timestamp string in the YYYY-MM-DDThh:mm:ss.sss.±hh:mm format, not {actual_type}.")]
+    InvalidThinEdgeJsonTime { actual_type: String },
+
+    #[error(
+        "Number out-of-range: the {name:?} value is too large to be represented as a float64."
+    )]
+    InvalidThinEdgeJsonNumber { name: String },
+
+    #[error("Invalid measurement name: {name:?} is a reserved word.")]
+    ThinEdgeReservedWordError { name: String },
+
+    #[error(
+        "Invalid ISO8601 timestamp (expected YYYY-MM-DDThh:mm:ss.sss.±hh:mm): {value:?}: {from}"
+    )]
+    InvalidTimestamp { value: String, from: ParseError },
+
+    #[error("More than 2 nested levels: the record for {name:?} must be flattened.")]
+    InvalidThinEdgeHierarchy { name: String },
+}
+
+impl ThinEdgeJsonError {
+    const MAX_LEN: usize = 80;
+
+    fn new_invalid_utf8(bytes: &[u8], from: std::str::Utf8Error) -> ThinEdgeJsonError {
+        let index = from.valid_up_to();
+        let input = std::str::from_utf8(&bytes[..index]).unwrap_or("");
+
+        ThinEdgeJsonError::InvalidUTF8 {
+            input_excerpt: input_prefix(input, ThinEdgeJsonError::MAX_LEN),
+            from,
+        }
+    }
+
+    fn new_invalid_json(input: &str, from: json::JsonError) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidJson {
+            input_excerpt: input_prefix(input, ThinEdgeJsonError::MAX_LEN),
+            from,
+        }
+    }
+
+    fn new_invalid_json_root(json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonRoot {
+            json_excerpt: input_prefix(&json.to_string(), ThinEdgeJsonError::MAX_LEN),
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn new_invalid_json_value(name: &str, json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonValue {
+            name: String::from(name),
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn new_invalid_json_time(json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonTime {
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn json_type(input: &JsonValue) -> &'static str {
+        match input {
+            JsonValue::String(_) | JsonValue::Short(_) => "a string",
+            JsonValue::Number(_) => "a number",
+            JsonValue::Object(_) => "an object",
+            JsonValue::Array(_) => "an array",
+            JsonValue::Boolean(_) => "a boolean",
+            JsonValue::Null => "null",
+        }
+    }
+}
```

Comment:
Missing new line and the end of the file.

Reply:
fixed